### ISSUE TITLE
Proposal: expose alternative query interfaces via `pim:storage`

### ISSUE THREAD

**Issue by rubensworks** [solid/specification#455]

**Comment by csarven** [comment]
Aren't the sparql, qpf etc interfaces about the storage and so should instead be part of storage's description?

**Comment by matthieubosquet** [comment]
SPARQL and QPF APIs could arguably be available over any resource (as opposed to just be a Pod-Wide `pim:storage` querying mechanism).

I think discovering such endpoints should happen via resource-based API discovery rather than via a WebID (which could contain multiple Solid storages).

I think that maybe this issue is most relevant to such requirements: https://github.com/solid/specification/issues/355.

**Comment by rubensworks** [comment]
> Aren't the sparql, qpf etc interfaces about the storage and so should instead be part of storage's description?

Indeed, that might make sense.

So the example from above based on your suggested storage description could become something like this:

```
<http://example.org/bob/container/>
  a pim:Storage .
  dcterms:title "okieli-dokieli LDP" .
  solid:owner <http://example.org/bob/#me> .


<http://example.org/bob/sparql/>
  a pim:Storage .
  dcterms:title "okieli-dokieli SPARQL" .
  solid:owner <http://example.org/bob/#me> .


<http://example.org/bob/qpf/>
  a pim:Storage .
  dcterms:title "okieli-dokieli QPF" .
  solid:owner <http://example.org/bob/#me> .
```

> SPARQL and QPF APIs could arguably be available over any resource (as opposed to just be a Pod-Wide pim:storage querying mechanism).
I think discovering such endpoints should happen via resource-based API discovery mechanism rather than via a WebID (which could contain multiple Solid storages).

Certainly resource-level interfaces definitely make sense. But I think the discovery of resource-level interfaces may work quite different to pod-level interfaces, so I'm inclined to focus purely on pod-level interfaces here.

**Comment by csarven** [comment]
It may not be appropriate to interpret those interfaces as another (type of a) storage. My understanding was that they are alternative (or additional) endpoints, services, communication options offered by the storage. Would a unique property indicating those interfaces suffice? Or employing VOID, SPARQL-SERVICE-DESCRIPTIONS, etc?

**Comment by rubensworks** [comment]
> It may not be appropriate to interpret those interfaces as another (type of a) storage. My understanding was that they are alternative (or additional) endpoints, services, communication options offered by the storage.

Yep, I agree with that view. With LDP also being one type of interface among these.

> Would a unique property indicating those interfaces suffice?

Is it something like this you have in mind?

```
<http://example.org/bob/container/>
  a pim:Storage ;
  dcterms:title "okieli-dokieli" ;
  solid:owner <http://example.org/bob/#me> ;
  solid:alternativeInterface <http://example.org/bob/sparql/>, <http://example.org/bob/qpf/>.
```

> Or employing VOID, SPARQL-SERVICE-DESCRIPTION, etc?

I think VOID and SPARQL SD definitely make sense to describe the capabilities of such interfaces, but we still need a way to discover these in the first place. In the examples above, `http://example.org/bob/sparql/` would expose the SPARQL SD.

**Comment by matthieubosquet** [comment]
> I agree, resource-level interfaces definitely make sense. But I think the discovery of resource-level interfaces may work quite different to pod-level interfaces, so I'm inclined to focus purely on pod-level interfaces here.

Maybe the discovery mechanism can be generic enough to accomodate advertising a SPARQL or qpf endpoint on any resource.

It seems to me that a [description resource](https://solidproject.org/TR/protocol#auxiliary-resources-description-resource) may be quite fitting for that purpose. It happens so that according to the [Solid Protocol ED](https://solidproject.org/ED/protocol#storage), servers MUST advertise a storage description resource (and maybe that storage description resource is the same as the resource one finds linked on the storage resource itself via the `link rel="describedby" header` or not, but they're both description resources).

Now, I think that there is a lot to be known about a [SPARQL endpoint](https://www.w3.org/TR/sparql11-service-description/#example-turtle) and that's why the SPARQL Service Description spec exists, we could probably efficiently leverage that, in the simplest manner by including a statement of the form `[] sd:endpoint <https://example.com/sparql>` in any description resource.

I don't mind introducing new vocabulary terms, but maybe it's nice to be more specific and have more expressivity when describing a service (qpf, sparql or solid).

**Comment by csarven** [comment]
Yes, along those lines. Perhaps the property is specialised for the interface so that clients can easily follow-their-nose.

I think `pim:Storage` is conceptually not equivalent to `void:Dataset` (or `dcat:Catalog`).

A Storage may want to share its VoID Datasets and DatasetDescriptions, e.g., discovered through their Description Resource (via `describedby` link relation or https://www.w3.org/ns/iana/link-relations/relation#describedby ).

Would `void:sparqlEndpoint` as a property of a void:Dataset suffice? Or do you think that these interfaces should really be about `pim:Storage`?

Is there something in LDF/QPF that indicates an interface or is that HYDRA (search/template?)

**Comment by rubensworks** [comment]
> Perhaps the property is specialised for the interface so that clients can easily follow-their-nose.

I don't think we necessarily _MUST_ have interface-specific properties, if we assume that interfaces are self-descriptive after dereferencing (which is the case within SPARQL endpoints via the SPARQL SD, and TPF/QPF via Hydra).
As long as you have the URL of an interface, this _should_ be sufficient. (for example, this is sufficient for Comunica)

However, specialised properties would make sense if we want to avoid clients having to dereference those interfaces before detecting their capabilities. But this might lead us quite far, perhaps even to the point that we will have to duplicate all information in the SPARQL SD and Hydra controls for TPF/QPF.

Furthermore, since such descriptions (SPARQL SD, Hydra, ...) usually remain static, they can be cached by clients, so I don't think this would involve too much processing overhead.

**Comment by elf-pavlik** [comment]
>> It may not be appropriate to interpret those interfaces as another (type of a) storage. My understanding was that they are alternative (or additional) endpoints, services, and communication options offered by the storage.

> Yep, I agree with that view. With LDP also being one type of interface among these.

:+1:

I think starting with something very generic like `void:Dataset` makes sense. Later adding more specific descriptions of interfaces for accessing it.
* Solid flavored LDP
* SPARQL
* QPF

etc.

No matter how the dataset is accessed, we are still dealing with the same set of quads.

**Comment by csarven** [comment]
>Solid flavored LDP

I don't believe that association is accurate and find it to be potentially misleading. The Solid Protocol and LDP are compatible in some aspects, however neither are an extension of the other. I suggest that we take care in how we communicate that. There are existing issues on this but we need to move some of the information towards the spec (or to another document.)

**Comment by RubenVerborgh** [comment]
`pim:storage` is a start, but it does not yet convey the "is an API to the same information" notion, so we'd need additional triples for that. I don't think it is useful without. I follow the above suggestions that compare to DCAT and others; I think the mechanism is similar to that of a dataset and a distribution.

I want to emphasize that the same data can be available through _multiple_ Solid Protocol interfaces (and even multiple QPFs, SPARQL endpoints, etc.), as we argue [here](https://solidlabresearch.github.io/WhatsInAPod/). Hence, just referring to them by type will not be enough (but can be a start).
